src/Databases/PostgreSQL/fetchPostgreSQLTableStructure.cpp

```diff
@@ -80,60 +94,120 @@ static DataTypePtr convertPostgreSQLDataType(std::string & type, bool is_nullabl
-           "WHERE attrelid = '{}'::regclass "
-           "AND NOT attisdropped AND attnum > 0", postgres_table_name);
     try
     {
-        pqxx::read_transaction tx(*connection);
-        pqxx::stream_from stream(tx, pqxx::from_query, std::string_view(query));
+        pqxx::stream_from stream(*tx, pqxx::from_query, std::string_view(query));
 
-        std::tuple<std::string, std::string, std::string, uint16_t> row;
-        while (stream >> row)
+        if (only_names_and_types)
         {
-            columns.push_back(NameAndTypePair(
-                    std::get<0>(row),
-                    convertPostgreSQLDataType(
-                        std::get<1>(row),
-                        use_nulls && (std::get<2>(row) == "f"), /// 'f' means that postgres `not_null` is false, i.e. value is nullable
-                        std::get<3>(row))));
+            std::tuple<std::string, std::string> row;
+            while (stream >> row)
+                columns.push_back(NameAndTypePair(std::get<0>(row), convertPostgreSQLDataType(std::get<1>(row))));
         }
+        else
+        {
+            std::tuple<std::string, std::string, std::string, uint16_t> row;
+            while (stream >> row)
+            {
+                columns.push_back(NameAndTypePair(
+                        std::get<0>(row),                           /// column name
+                        convertPostgreSQLDataType(
+                            std::get<1>(row),                       /// data type
+                            use_nulls && (std::get<2>(row) == "f"), /// 'f' means that postgres `not_null` is false
+                            std::get<3>(row))));                    /// number of dimensions if data type is array
+            }
+        }
+
         stream.complete();
-        tx.commit();
     }
     catch (const pqxx::undefined_table &)
     {
         throw Exception(fmt::format(
-                    "PostgreSQL table {}.{} does not exist",
-                    connection->dbname(), postgres_table_name), ErrorCodes::UNKNOWN_TABLE);
+                    "PostgreSQL table {} does not exist", postgres_table_name), ErrorCodes::UNKNOWN_TABLE);
     }
     catch (Exception & e)
     {
         e.addMessage("while fetching postgresql table structure");
         throw;
     }
 
-    if (columns.empty())
-        return nullptr;
+    return !columns.empty() ? std::make_shared<NamesAndTypesList>(columns) : nullptr;
+}
+
+
+template<typename T>
+PostgreSQLTableStructure fetchPostgreSQLTableStructureImpl(
+        std::shared_ptr<T> tx, const String & postgres_table_name, bool use_nulls, bool with_primary_key)
+{
+    PostgreSQLTableStructure table;
+
+    if (postgres_table_name.find('\'') != std::string::npos
+        || postgres_table_name.find('\\') != std::string::npos)
+    {
+        throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "PostgreSQL table name cannot contain single quote or backslash characters, passed {}",
```

Comment:
There was a different reason why this was added here https://github.com/ClickHouse/ClickHouse/pull/18554/commits/b1b09ed5ae5ede2609de4a882e94a565a5e68b14

Reply:
We have multiple functions to write identifiers with different quoting styles (see `IdentifierQuotingStyle` and functions like `backQuoteIfNeed(...)`, `writeBackQuotedStringMySQL(...)`, etc). Maybe use one of them (or add new quoting style for Postgres if existing ones are not suitable)? It will protest from injection too, but also will support all possible names. You can do it in a separate PR if you want to.

Reply:
> Maybe use one of them

Done.